jepsen/src/jepsen/client.clj

```diff
@@ -1,20 +1,69 @@
 (ns jepsen.client
-  "Applies operations to a database.")
+  "Applies operations to a database."
+  (:require [clojure.tools.logging :refer :all]
+            [clojure.reflect :refer [reflect]]
+            [jepsen.util :as util]))
 
 (defprotocol Client
-  (setup!  [client test node]
+  (open! [client test node]
           "Set up the client to work with a particular node. Returns a client
           which is ready to accept operations via invoke!")
+  (close! [client test node]
+          "Close the client connection when work is completed or an invocation
+           crashes the client.")
+  (setup! [client test] [client test node]
+          "Called once to set up database state for testing. 3 arity form is
+           deprecated and will be removed in a future jepsen version.")
   (invoke! [client test operation]
            "Apply an operation to the client, returning an operation to be
            appended to the history. For multi-stage operations, the client may
            reach into the test and conj onto the history atom directly.")
   (teardown! [client test]
-             "Tear down the client when work is complete."))
+           "Tear down the client when work is complete."))
 
 (def noop
   "Does nothing."
   (reify Client
-    (setup!    [this test node] this)
+    (setup!    [this test])
     (teardown! [this test])
-    (invoke!   [this test op] (assoc op :type :ok))))
+    (invoke!   [this test op] (assoc op :type :ok))
+    (open!     [this test node] this)
+    (close!    [this test node])))
+
+(defn reopen!
+  "Takes the worker's current client, closes its connection and
+  opens a new connection with the original."
```

Comment:
The docstring says this opens a new connection with the original, but it looks like it opens the current client again. Maybe... drop this function and use close!/open! directly in jepsen.core?

Reply:
Yeah, this particular abstraction didn't win us much. It's been excised